src/libraries/System.Private.CoreLib/src/System/Threading/Overlapped.cs

```diff
@@ -0,0 +1,231 @@
+                }
+
+                pNativeOverlapped = (NativeOverlapped*)NativeMemory.Alloc(
+                    (nuint)(sizeof(NativeOverlapped) + sizeof(nuint)) + handleCount * (nuint)sizeof(GCHandle));
+
+                GCHandleCountRef(pNativeOverlapped) = 0;
+
+                pNativeOverlapped->InternalLow = default;
+                pNativeOverlapped->InternalHigh = default;
+                pNativeOverlapped->OffsetLow = _offsetLow;
+                pNativeOverlapped->OffsetHigh = _offsetHigh;
+                pNativeOverlapped->EventHandle = _eventHandle;
+
+                GCHandleRef(pNativeOverlapped, 0) = GCHandle.Alloc(this);
+                GCHandleCountRef(pNativeOverlapped)++;
+
+                if (userData != null)
+                {
+                    if (userData.GetType() == typeof(object[]))
+                    {
+                        object[] objArray = (object[])userData;
+                        for (int i = 0; i < objArray.Length; i++)
+                        {
+                            GCHandleRef(pNativeOverlapped, (nuint)(i + 1)) = GCHandle.Alloc(objArray[i], GCHandleType.Pinned);
+                            GCHandleCountRef(pNativeOverlapped)++;
+                        }
+                    }
+                    else
+                    {
+                        GCHandleRef(pNativeOverlapped, 1) = GCHandle.Alloc(userData, GCHandleType.Pinned);
+                        GCHandleCountRef(pNativeOverlapped)++;
+                    }
+                }
+
+                Debug.Assert(GCHandleCountRef(pNativeOverlapped) == handleCount);
+
+#if FEATURE_PERFTRACING
+#if !(TARGET_BROWSER && !FEATURE_WASM_THREADS)
+                if (NativeRuntimeEventSource.Log.IsEnabled())
+                    NativeRuntimeEventSource.Log.ThreadPoolIOPack(pNativeOverlapped);
+#endif
+#endif
+
+                NativeOverlapped* pRet = _pNativeOverlapped = pNativeOverlapped;
+                pNativeOverlapped = null;
+                return pRet;
+            }
+            finally
+            {
+                if (pNativeOverlapped != null)
+                    FreeNativeOverlapped(pNativeOverlapped);
+            }
+        }
+
+        internal static void FreeNativeOverlapped(NativeOverlapped* pNativeOverlapped)
+        {
+            nuint handleCount = GCHandleCountRef(pNativeOverlapped);
+
+            for (nuint i = 0; i < handleCount; i++)
+                GCHandleRef(pNativeOverlapped, i).Free();
+
+            NativeMemory.Free(pNativeOverlapped);
+        }
+
+        //
+        // The NativeOverlapped structure is followed by GC handle count and inline array of GC handles
+        //
+        private static ref nuint GCHandleCountRef(NativeOverlapped* pNativeOverlapped)
+            => ref *(nuint*)(pNativeOverlapped + 1);
+
+        private static ref GCHandle GCHandleRef(NativeOverlapped* pNativeOverlapped, nuint index)
+            => ref *((GCHandle*)((nuint*)(pNativeOverlapped + 1) + 1) + index);
```

Comment:
We have enough features for C# to look like C++ 😄 

![Rpx_syriaMicDropGIF](https://user-images.githubusercontent.com/95136/186569505-f8f20fd4-1c8b-46e2-aa99-089e66e20ae3.gif)

Reply:
Gotta remember that right-left rule.